README.md
This is MATE Tweak, a fork of [mintDesktop](https://github.com/linuxmint/mintdesktop).

  * MATE Tweak removes the Mint specific configuration options.

Personally I'm not the least bit interested in using the MATE Tweak but I 
see that it is regularly requested in the Ubuntu MATE community. So 
consider MATE Tweak a gift from me, to you :-)

Anyone interested in testing MATE Tweak I've made packages for Trusty and
Utopic, which you can install from the following PPA:

  * https://launchpad.net/~ubuntu-mate-dev/+archive/ubuntu/ppa


debian/changelog
mate-tweak (3.4.1-0build1) UNRELEASED; urgency=medium

  * Switch into non-native package.

 -- Mike Gabriel <sunweaver@debian.org>  Wed, 18 Feb 2015 14:07:39 +0100

mate-tweak (3.4.1) vivid; urgency=medium

  * Synced with mintDesktop.
  * Added Compiz support with live switching between Marco and Compiz.
  * Added panel layout interface switcher.
  * Change button-layout for MATE, GTK3 and GNOME.

 -- Martin Wimpress <code@flexion.org>  Sun, 18 Jan 2015 21:54:22 +0000

mate-tweak (3.3.6) vivid; urgency=medium

  * Initial fork of mintDesktop.
  * Removed configuration options specific to Linux Mint.

 -- Martin Wimpress <code@ubuntu-mate.org>  Sun, 02 Nov 2014 00:50:40 +0100
 
mintdesktop (3.3.6) qiana; urgency=medium

  * Simplified UI
  * Switched to GTK3
  * Enabled marco detection for windows tab

 -- Clement Lefebvre <root@linuxmint.com>  Tue, 06 May 2014 12:41:41 +0100

mintdesktop (3.3.5) qiana; urgency=medium

  * Updated desktop file

 -- Clement Lefebvre <root@linuxmint.com>  Sun, 04 May 2014 02:48:31 +0100

mintdesktop (3.3.4) qiana; urgency=medium

  * Removed temp file from package (was added by mistake)

 -- Clement Lefebvre <root@linuxmint.com>  Fri, 02 May 2014 12:58:19 +0100

mintdesktop (3.3.3) petra; urgency=low

  * Show mintfortune conf (systemd fixed)

 -- Clement Lefebvre <root@linuxmint.com>  Fri, 22 Nov 2013 10:45:52 +0000

mintdesktop (3.3.2) petra; urgency=low

  * Hide mintfortune conf for now (removed from mintsystem until systemd/XDG_RUNTIME_DIR is fixed)

 -- Clement Lefebvre <root@linuxmint.com>  Mon, 11 Nov 2013 18:29:34 +0000

mintdesktop (3.3.1) olivia; urgency=low

  * Migrated to MATE 1.6 (gsettings, gi, etc..)

 -- Clement Lefebvre <root@linuxmint.com>  Wed, 13 Mar 2013 13:52:40 +0000

mintdesktop (3.3.0) olivia; urgency=low

  * Changed description in debian/control

 -- Clement Lefebvre <root@linuxmint.com>  Tue, 12 Mar 2013 15:36:29 +0000

mintdesktop (3.2.9) nadia; urgency=low

  * Removed about dialog (useless)
  * Changed "Use GNOME Compositing" label to "Use Compositing" (refers to Marco's)
  * Added cinnamon in the list of known desktop sessions
  * Updated POT File

 -- Clement Lefebvre <root@linuxmint.com>  Sun, 04 Nov 2012 10:52:10 +0000

mintdesktop (3.2.8) maya; urgency=low

  * Fixed wrong icon name

 -- Clement Lefebvre <root@linuxmint.com>  Mon, 07 May 2012 16:50:07 +0100

mintdesktop (3.2.7) maya; urgency=low

  * Dedicated mintDesktop to MATE
  * Fixed session detection in MATE
  * Fixed menu file mentioning Gnome

 -- Clement Lefebvre <root@linuxmint.com>  Wed, 18 Apr 2012 12:38:42 +0100

mintdesktop (3.2.6) lisa; urgency=low

  * Added full support for MATE

 -- Clement Lefebvre <root@linuxmint.com>  Tue,  8 Nov 2011 10:31:00 +0000

mintdesktop (3.2.5) lisa; urgency=low

  * Different panels for mate and gnome3

 -- Clement Lefebvre <root@linuxmint.com>  Fri,  28 Oct 2011 11:59:00 +0000

mintdesktop (3.2.4) katya; urgency=low

  * Katya  

 -- Clement Lefebvre <root@linuxmint.com>  Tue,  12 Apr 2011 09:22:00 +0000

mintdesktop (3.2.3) julia; urgency=low

  * Better looking UI
  * Added terminal option to hide/show fortune cookies

 -- Clement Lefebvre <root@linuxmint.com>  Fri,  18 Feb 2011 12:12:00 +0000

mintdesktop (3.2.2) julia; urgency=low

  * Removed obsolete/buggy prerm

 -- Clement Lefebvre <root@linuxmint.com>  Tue,  05 Oct 2010 15:46:00 +0000

mintdesktop (3.2.1) isadora; urgency=low

  * Updated desktop files

 -- Clement Lefebvre <root@linuxmint.com>  Fri,  14 May 2010 10:32:00 +0000

mintdesktop (3.2.0) isadora; urgency=low

  * Renamed button layouts

 -- Clement Lefebvre <root@linuxmint.com>  Wed,  28 Apr 2010 14:34:00 +0000

mintdesktop (3.1.9) isadora; urgency=low

  * Renamed button layouts

 -- Clement Lefebvre <root@linuxmint.com>  Sun,  25 Apr 2010 19:41:00 +0000

mintdesktop (3.1.8) isadora; urgency=low

  * Ikey changes
  * i18n
  * Relooking

 -- Clement Lefebvre <root@linuxmint.com>  Sun,  18 Apr 2010 18:03:00 +0000

mintdesktop (3.1.7) isadora; urgency=low

  * Updated menu button styles

 -- Clement Lefebvre <root@linuxmint.com>  Sun,  28 Mar 2010 13:20:00 +0000

mintdesktop (3.1.6) isadora; urgency=low

  * Added window buttons layout styles
  * Changes are now immediate, no need for an apply button

 -- Clement Lefebvre <root@linuxmint.com>  Sun,  28 Mar 2010 12:16:00 +0000

mintdesktop (3.1.5) helena; urgency=low

  * Changed translation framework

 -- Clement Lefebvre <root@linuxmint.com>  Tue,  3 Nov 2009 11:57:00 +0000

mintdesktop (3.1.4) helena; urgency=low

  * Blueprint implemented: https://blueprints.launchpad.net/linuxmint/+spec/generic-branding

 -- Clement Lefebvre <root@linuxmint.com>  Mon,  27 Jul 2009 18:12:00 +0000

mintdesktop (3.1.3) gloria; urgency=low

  * Updated translations

 -- Clement Lefebvre <root@linuxmint.com>  Tue,  19 May 2009 18:55:00 +0000

mintdesktop (3.1.2) gloria; urgency=low

  * Updated translations

 -- Clement Lefebvre <root@linuxmint.com>  Sun,  17 May 2009 19:58:00 +0000

mintdesktop (3.1.1) gloria; urgency=low

  * Added translations

 -- Clement Lefebvre <root@linuxmint.com>  Sun,  17 May 2009 18:07:00 +0000

mintdesktop (3.1) gloria; urgency=low

  * Removed splash screen management

usr/lib/linuxmint/mintDesktop/mintDesktop.py
#!/usr/bin/env python

import gi
 
from gi.repository import Gtk, GdkPixbuf, Gdk, GObject
from gi.repository import Gio

try:
    import os
    import commands
    import sys
    import string
    import gettext    
except Exception, detail:
    print detail
    sys.exit(1)


# i18n
# TODO: Badly need to fix this - overuse of "The" etc.
gettext.install("mintdesktop", "/usr/share/linuxmint/locale")

class SidePage:
    def __init__(self, notebook_index, name, icon):
        self.notebook_index = notebook_index
        self.name = name
        self.icon = icon

class MintDesktop:

    def set_string(self, schema, key, value):
        settings = Gio.Settings.new(schema)
        settings.set_string(key, value)

    def get_string(self, schema, key):
        settings = Gio.Settings.new(schema)
        return settings.get_string(key)

    def set_bool(self, schema, key, value):
        settings = Gio.Settings.new(schema)
        settings.set_boolean(key, value.get_active())        
    
    def get_bool(self, schema, key):
        settings = Gio.Settings.new(schema)
        return settings.get_boolean(key)
    
    def init_checkbox(self, schema, key, name):
        widget = self.builder.get_object(name)
        value = self.get_bool(schema, key)
        widget.set_active(value)
        widget.connect("clicked", lambda x: self.set_bool(schema, key, x))       

    def init_combobox(self, schema, key, name):
        widget = self.builder.get_object(name)
        conf = self.get_string(schema, key)
        index = 0
        for row in widget.get_model():
            if(conf == row[1]):
                widget.set_active(index)
                break
            index = index +1
        widget.connect("changed", lambda x: self.combo_fallback(schema, key, x))

    def combo_fallback(self, schema, key, widget):
        act = widget.get_active()
        value = widget.get_model()[act]
        self.set_string(schema, key, value[1])

    # Change pages
    def side_view_nav(self, param):
        treePaths = param.get_selected_items()        
        if (len(treePaths) > 0):
            treePath = treePaths[0]
            index = int("%s" % treePath) #Hack to turn treePath into an int
            target = self.sidePages[index].notebook_index
            self.builder.get_object("notebook1").set_current_page(index)

    ''' Create the UI '''
    def __init__(self):
        # Detect which WM is running
        marco_mode = False
        wm_info = commands.getoutput("wmctrl -m")        
        if "Marco" in wm_info:
            print "Marco WM detected"
            marco_mode = True

        # load our glade ui file in
        self.builder = Gtk.Builder()
        self.builder.add_from_file('/usr/lib/linuxmint/mintDesktop/mintDesktop.ui')
        self.window = self.builder.get_object( "main_window" )
               
        self.builder.get_object("main_window").connect("destroy", Gtk.main_quit)
                      
        side_desktop_options = SidePage(0, _("Desktop"), "user-desktop")
        side_windows = SidePage(1, _("Windows"), "preferences-system-windows")
        side_interface = SidePage(2, _("Interface"), "preferences-desktop")
        side_terminal = SidePage(3, _("Terminal"), "terminal")
                
        if marco_mode:
            self.sidePages = [side_desktop_options, side_windows, side_interface, side_terminal]
        else:
            self.sidePages = [side_desktop_options, side_interface, side_terminal]
                
        # create the backing store for the side nav-view.                    
        theme = Gtk.IconTheme.get_default()
        self.store = Gtk.ListStore(str, GdkPixbuf.Pixbuf)
        for sidePage in self.sidePages:
            img = theme.load_icon(sidePage.icon, 36, 0)                        
            self.store.append([sidePage.name, img])       
            
        target = self.sidePages[0].notebook_index
        self.builder.get_object("notebook1").set_current_page(target)                                            

        # set up the side view - navigation.
        self.builder.get_object("side_view").set_text_column(0)
        self.builder.get_object("side_view").set_pixbuf_column(1)
        self.builder.get_object("side_view").set_model(self.store)
        self.builder.get_object("side_view").select_path(Gtk.TreePath.new_first())
        self.builder.get_object("side_view").connect("selection_changed", self.side_view_nav )

        # set up larger components.
        self.builder.get_object("main_window").set_title("Desktop Settings")
        self.builder.get_object("main_window").connect("destroy", Gtk.main_quit)

        # i18n
        self.builder.get_object("label_desktop_icons").set_markup("<b>" + _("Desktop icons") + "</b>")
        self.builder.get_object("label_performance").set_markup("<b>" + _("Performance") + "</b>")
        self.builder.get_object("label_appearance").set_markup("<b>" + _("Appearance") + "</b>")
        self.builder.get_object("label_icons").set_markup("<b>" + _("Icons") + "</b>")
        self.builder.get_object("label_context_menus").set_markup("<b>" + _("Context menus") + "</b>")
        self.builder.get_object("label_toolbars").set_markup("<b>" + _("Toolbars") + "</b>")
        self.builder.get_object("label_terminal").set_markup("<b>" + _("Terminal") + "</b>")

        self.builder.get_object("caption_desktop_icons").set_markup("<small><i><span foreground=\"#555555\">" + _("Select the items you want to see on the desktop:") + "</span></i></small>")

        self.builder.get_object("checkbox_computer").set_label(_("Computer"))
        self.builder.get_object("checkbox_home").set_label(_("Home"))
        self.builder.get_object("checkbox_network").set_label(_("Network"))
        self.builder.get_object("checkbox_trash").set_label(_("Trash"))
        self.builder.get_object("checkbox_volumes").set_label(_("Mounted Volumes"))

        self.builder.get_object("checkbutton_resources").set_label(_("Don't show window content while dragging them"))
        self.builder.get_object("checkbox_compositing").set_label(_("Use compositing"))
        self.builder.get_object("checkbutton_titlebar").set_label(_("Use system font in titlebar"))
        self.builder.get_object("checkbox_fortunes").set_label(_("Show fortune cookies"))

        self.builder.get_object("label_layouts").set_text(_("Buttons layout:"))

        self.builder.get_object("checkbutton_menuicon").set_label(_("Show icons on menus"))
        self.builder.get_object("checkbutton_button_icons").set_label(_("Show icons on buttons"))
        self.builder.get_object("checkbutton_im_menu").set_label(_("Show Input Methods menu in context menus"))
        self.builder.get_object("checkbutton_unicode").set_label(_("Show Unicode Control Character menu in context menus"))

        self.builder.get_object("label_tool_icons").set_text(_("Buttons labels:"))
        self.builder.get_object("label_icon_size").set_text(_("Icon size:"))

        # Desktop page
        self.init_checkbox("org.mate.caja.desktop", "computer-icon-visible", "checkbox_computer")
        self.init_checkbox("org.mate.caja.desktop", "home-icon-visible", "checkbox_home")
        self.init_checkbox("org.mate.caja.desktop", "network-icon-visible", "checkbox_network")
        self.init_checkbox("org.mate.caja.desktop", "trash-icon-visible", "checkbox_trash")
        self.init_checkbox("org.mate.caja.desktop", "volumes-visible", "checkbox_volumes")

        # Window Manager page
        self.init_checkbox("org.mate.Marco.general", "reduced-resources", "checkbutton_resources")
        self.init_checkbox("org.mate.Marco.general", "compositing-manager", "checkbox_compositing")
        self.init_checkbox("org.mate.Marco.general", "titlebar-uses-system-font", "checkbutton_titlebar")

        # interface page
        self.init_checkbox("org.mate.interface", "menus-have-icons", "checkbutton_menuicon")
        self.init_checkbox("org.mate.interface", "show-input-method-menu","checkbutton_im_menu")
        self.init_checkbox("org.mate.interface", "show-unicode-menu", "checkbutton_unicode")
        self.init_checkbox("org.mate.interface", "buttons-have-icons", "checkbutton_button_icons")
        
        # terminal page
        self.init_checkbox("com.linuxmint.terminal", "show-fortunes", "checkbox_fortunes")

        iconSizes = Gtk.ListStore(str, str)
        iconSizes.append([_("Small"), "small-toolbar"])
        iconSizes.append([_("Large"), "large-toolbar"])
        self.builder.get_object("combobox_icon_size").set_model(iconSizes)
        self.init_combobox("org.mate.interface", "toolbar-icons-size", "combobox_icon_size")

        # Metacity button layouts..
        layouts = Gtk.ListStore(str, str)
        layouts.append([_("Traditional style (Right)"), "menu:minimize,maximize,close"])
        layouts.append([_("Mac style (Left)"), "close,minimize,maximize:"])
        self.builder.get_object("combo_wmlayout").set_model(layouts)
        self.init_combobox("org.mate.Marco.general", "button-layout", "combo_wmlayout")

        # toolbar icon styles
        iconStyles = Gtk.ListStore(str, str)
        iconStyles.append([_("Text below items"), "both"])
        iconStyles.append([_("Text beside items"), "both-horiz"])
        iconStyles.append([_("Icons only"), "icons"])
        iconStyles.append([_("Text only"), "text"])
        self.builder.get_object("combobox_toolicons").set_model(iconStyles)
        self.init_combobox("org.mate.interface", "toolbar-style", "combobox_toolicons")

        self.builder.get_object("main_window").show()       

    
if __name__ == "__main__":
    MintDesktop()
    Gtk.main()


usr/lib/ubuntu-mate/mate-tweak/mate-tweak.py
#!/usr/bin/env python

import gi
import os
import shutil
import string
import sys
import time

from gi.repository import Gtk, GdkPixbuf, Gdk, GObject
from gi.repository import Gio
from subprocess import Popen

try:
    import commands
    import gettext
except Exception, detail:
    print detail
    sys.exit(1)

# i18n
# TODO: Badly need to fix this - overuse of "The" etc.
gettext.install("mate-tweak", "/usr/share/ubuntu-mate/locale")

class SidePage:
    def __init__(self, notebook_index, name, icon):
        self.notebook_index = notebook_index
        self.name = name
        self.icon = icon

class MateTweak:

    def set_string(self, schema, key, value):
        settings = Gio.Settings.new(schema)
        settings.set_string(key, value)

    def get_string(self, schema, key):
        settings = Gio.Settings.new(schema)
        return settings.get_string(key)

    def set_bool(self, schema, key, value):
        settings = Gio.Settings.new(schema)
        settings.set_boolean(key, value.get_active())
    
    def get_bool(self, schema, key):
        settings = Gio.Settings.new(schema)
        return settings.get_boolean(key)
    
    def init_checkbox(self, schema, key, name):
        source = Gio.SettingsSchemaSource.get_default()
        if source.lookup(schema, True) != None:
            widget = self.builder.get_object(name)
            value = self.get_bool(schema, key)
            widget.set_active(value)
            widget.connect("clicked", lambda x: self.set_bool(schema, key, x))       

    def init_combobox(self, schema, key, name):
        source = Gio.SettingsSchemaSource.get_default()
        if source.lookup(schema, True) != None:
            widget = self.builder.get_object(name)
            conf = self.get_string(schema, key)
            index = 0
            for row in widget.get_model():
                if(conf == row[1]):
                    widget.set_active(index)                    
                    break
                index = index +1
            widget.connect("changed", lambda x: self.combo_fallback(schema, key, x))

    def find_on_path(self, command):
        """Is command on the executable search path?"""
        if 'PATH' not in os.environ:
            return False
        path = os.environ['PATH']
        for element in path.split(os.pathsep):
            if not element:
                continue
            filename = os.path.join(element, command)
            if os.path.isfile(filename) and os.access(filename, os.X_OK):
                return True
        return False

    def replace_windowmanager(self, wm):
        # Use this in python < 3.3. Python >= 3.3 has subprocess.DEVNULL
        devnull = open(os.devnull, 'wb')
        if wm == 'compiz':
            # mate-panel needs to be replaced when switching to Compiz to avoid
            # rendering screw ups.
            Popen(['killall', 'mate-panel'], stdout=devnull, stderr=devnull)
            Popen([wm, '--replace', 'ccp'], stdout=devnull, stderr=devnull)
            Popen(['mate-panel', '--replace'], stdout=devnull, stderr=devnull)
        else:
            Popen([wm, '--replace'], stdout=devnull, stderr=devnull)
        devnull.close()

    def enable_dock(self, dock):
        shutil.copy2('/usr/share/applications/' + dock + '.desktop', os.path.expanduser('~/.config/autostart/'))
		
        devnull = open(os.devnull, 'wb')
		# Docks require compositing
        current_wm = commands.getoutput("wmctrl -m")
        if "Marco" in current_wm:		
            settings = Gio.Settings.new('org.mate.Marco.general')
            settings.set_boolean('compositing-manager', True)        
            Popen(['marco', '--replace'], stdout=devnull, stderr=devnull)        
        Popen([dock], stdout=devnull, stderr=devnull) 
        devnull.close()

    def disable_dock(self, dock):		
        if os.path.exists(os.path.expanduser('~/.config/autostart/') + dock + '.desktop'):
            os.remove(os.path.expanduser('~/.config/autostart/') + dock + '.desktop')
        devnull = open(os.devnull, 'wb')
        Popen(['killall', dock], stdout=devnull, stderr=devnull)         
        devnull.close()

    def enable_mate_volume_applet(self):
        devnull = open(os.devnull, 'wb')		
        Popen(['mate-volume-control-applet'], stdout=devnull, stderr=devnull)		
        devnull.close()
        
        # If the system autostart is missing create a user autostart.
        if not os.path.exists('/etc/xdg/autostart/mate-volume-control-applet.desktop'):
            shutil.copy2('/usr/lib/ubuntu-mate/mate-tweak/mate-volume-control-applet.desktop', os.path.expanduser('~/.config/autostart/'))			
        	
    def disable_mate_volume_applet(self):
        devnull = open(os.devnull, 'wb')        
        Popen(['killall', 'mate-volume-control-applet'], stdout=devnull, stderr=devnull)
        
        if os.path.exists(os.path.expanduser('~/.config/autostart/') + 'mate-volume-control-applet.desktop'):
            os.remove(os.path.expanduser('~/.config/autostart/') + 'mate-volume-control-applet.desktop')
        if os.path.exists('/etc/xdg/autostart/mate-volume-control-applet.desktop'):
			Popen(['pkexec', '/usr/lib/ubuntu-mate/mate-tweak/disable-mate-volume-applet'], stdout=devnull, stderr=devnull)
        devnull.close()
        
    def replace_panel_layout(self, panel_layout):		
        # Use this in python < 3.3. Python >= 3.3 has subprocess.DEVNULL        
        devnull = open(os.devnull, 'wb')       
        Popen(['killall', 'mate-panel'], stdout=devnull, stderr=devnull)
        os.system('mate-panel --layout ' + panel_layout + ' --reset')
        Popen(['mate-panel', '--replace'], stdout=devnull, stderr=devnull)
        devnull.close()
        
        # Indicator enabled layouts auto-load the sound-indicator.
        # Therefore, make mate-volume-control-applet is killed or 
        # started accordingly.
        if 'indicators' in panel_layout:
            self.disable_mate_volume_applet()
        else:
            self.enable_mate_volume_applet()
        
        # Enable/disable the appropriate dock if one of the dock ready
        # panel layouts is selected.
        dock = None
        if os.path.exists('/usr/bin/docky'):
            dock = 'docky'	
        elif os.path.exists('/usr/bin/plank'):
            dock = 'plank'							
                    
        if panel_layout.startswith('eleven') and dock is not None:			
            if dock is 'docky':
                self.disable_dock('plank')
            elif dock is 'plank':
                self.disable_dock('docky')

            self.enable_dock(dock)
        else:
            self.disable_dock('docky')
            self.disable_dock('plank')

    def panel_layout_exists(self, panel_layout):
        if os.path.exists('/usr/share/mate-panel/layouts/' + panel_layout + '.layout'):
            return True
        return False

    def additional_tweaks(self, schema, key, value):
        if schema == "org.mate.Marco.general" and key == "button-layout":
            # If the button-layout is changed in MATE reflect that change
            # for GTK3 and GNOME.
            self.set_string("org.mate.interface", "gtk-decoration-layout", value)
            self.set_string("org.gnome.desktop.wm.preferences", "button-layout", value)

        elif schema == "org.mate.session.required-components" and key == "windowmanager":
            wm = value
            # Sanity check
            if wm not in ['compiz', 'marco']:
                wm = 'marco'

            # If the window manager is being changed, replace it now!
            print('Replacing the window manager with ' + wm)
            self.replace_windowmanager(wm)
            
            # As we are replacing the window manager, exit MATE Tweak.
            sys.exit(0)            
            
        elif schema == "org.mate.panel" and key == "default-layout":            
            panel_layout = value
            # If the panel layout is being changed, replace it now!
            self.replace_panel_layout(panel_layout)

    def combo_fallback(self, schema, key, widget):
        act = widget.get_active()
        value = widget.get_model()[act]
        self.set_string(schema, key, value[1])

        # Process any additional changes required for the schema and key
        self.additional_tweaks(schema, key, value[1])

    # Change pages
    def side_view_nav(self, param):
        treePaths = param.get_selected_items()
        if (len(treePaths) > 0):
            treePath = treePaths[0]
            index = int("%s" % treePath) #Hack to turn treePath into an int
            target = self.sidePages[index].notebook_index
            self.builder.get_object("notebook1").set_current_page(target)

    ''' Create the UI '''
    def __init__(self):
        # load our glade ui file in
        self.builder = Gtk.Builder()
        self.builder.add_from_file('/usr/lib/ubuntu-mate/mate-tweak/mate-tweak.ui')
        self.window = self.builder.get_object( "main_window" )
               
        self.builder.get_object("main_window").connect("destroy", Gtk.main_quit)
                      
        side_desktop_options = SidePage(0, _("Desktop"), "user-desktop")
        side_windows = SidePage(1, _("Windows"), "preferences-system-windows")
        side_interface = SidePage(2, _("Interface"), "preferences-desktop")
        
        # Determine the currently active window manager
        marco_mode = False
        compiz_mode = False        
        current_wm = commands.getoutput("wmctrl -m")
        if "Marco" in current_wm:
            marco_mode = True
        elif "Compiz" in current_wm:
            compiz_mode = True
        
        # Do not show Marco configuration options because Marco is not
        # currently running.
        if not marco_mode:
            self.builder.get_object("frame_marco1").hide()

        if marco_mode or compiz_mode:
            # Show window manager configuration
            self.sidePages = [side_desktop_options, side_interface, side_windows]
        else:
            # Do not show window manager configuration because the
            # currently running window manager is unknown.
            self.sidePages = [side_desktop_options, side_interface]
            self.builder.get_object("frame_marco2").hide()

        # create the backing store for the side nav-view.
        theme = Gtk.IconTheme.get_default()
        self.store = Gtk.ListStore(str, GdkPixbuf.Pixbuf)
        for sidePage in self.sidePages:
            img = theme.load_icon(sidePage.icon, 36, 0)
            self.store.append([sidePage.name, img])

        target = self.sidePages[0].notebook_index
        self.builder.get_object("notebook1").set_current_page(target)

        # set up the side view - navigation.
        self.builder.get_object("side_view").set_text_column(0)
        self.builder.get_object("side_view").set_pixbuf_column(1)
        self.builder.get_object("side_view").set_model(self.store)
        self.builder.get_object("side_view").select_path(Gtk.TreePath.new_first())
        self.builder.get_object("side_view").connect("selection_changed", self.side_view_nav )

        # set up larger components.
        self.builder.get_object("main_window").set_title("MATE Tweak")
        self.builder.get_object("main_window").connect("destroy", Gtk.main_quit)

        # i18n
        self.builder.get_object("label_desktop_icons").set_markup("<b>" + _("Desktop icons") + "</b>")
        self.builder.get_object("label_performance").set_markup("<b>" + _("Performance") + "</b>")
        self.builder.get_object("label_appearance").set_markup("<b>" + _("Appearance") + "</b>")
        self.builder.get_object("label_icons").set_markup("<b>" + _("Icons") + "</b>")
        self.builder.get_object("label_context_menus").set_markup("<b>" + _("Context menus") + "</b>")
        self.builder.get_object("label_toolbars").set_markup("<b>" + _("Toolbars") + "</b>")
        self.builder.get_object("label_wm").set_markup("<b>" + _("Window Manager") + "</b>")

        self.builder.get_object("caption_desktop_icons").set_markup("<small><i><span foreground=\"#555555\">" + _("Select the items you want to see on the desktop:") + "</span></i></small>")

        self.builder.get_object("checkbox_computer").set_label(_("Computer"))
        self.builder.get_object("checkbox_home").set_label(_("Home"))
        self.builder.get_object("checkbox_network").set_label(_("Network"))
        self.builder.get_object("checkbox_trash").set_label(_("Trash"))
        self.builder.get_object("checkbox_volumes").set_label(_("Mounted Volumes"))

        self.builder.get_object("checkbutton_resources").set_label(_("Don't show window content while dragging them"))
        self.builder.get_object("checkbox_compositing").set_label(_("Use compositing"))

        self.builder.get_object("label_layouts").set_text(_("Buttons layout:"))
        self.builder.get_object("label_window_manager").set_text(_("Window manager:"))

        self.builder.get_object("checkbutton_menuicon").set_label(_("Show icons on menus"))
        self.builder.get_object("checkbutton_button_icons").set_label(_("Show icons on buttons"))
        self.builder.get_object("checkbutton_im_menu").set_label(_("Show Input Methods menu in context menus"))
        self.builder.get_object("checkbutton_unicode").set_label(_("Show Unicode Control Character menu in context menus"))

        self.builder.get_object("label_tool_icons").set_text(_("Buttons labels:"))
        self.builder.get_object("label_icon_size").set_text(_("Icon size:"))

        # Desktop page
        self.init_checkbox("org.mate.caja.desktop", "computer-icon-visible", "checkbox_computer")
        self.init_checkbox("org.mate.caja.desktop", "home-icon-visible", "checkbox_home")
        self.init_checkbox("org.mate.caja.desktop", "network-icon-visible", "checkbox_network")
        self.init_checkbox("org.mate.caja.desktop", "trash-icon-visible", "checkbox_trash")
        self.init_checkbox("org.mate.caja.desktop", "volumes-visible", "checkbox_volumes")

        # Window Manager page        
        self.init_checkbox("org.mate.Marco.general", "reduced-resources", "checkbutton_resources")
        self.init_checkbox("org.mate.Marco.general", "compositing-manager", "checkbox_compositing")

        # interface page
        self.init_checkbox("org.mate.interface", "menus-have-icons", "checkbutton_menuicon")
        self.init_checkbox("org.mate.interface", "show-input-method-menu","checkbutton_im_menu")
        self.init_checkbox("org.mate.interface", "show-unicode-menu", "checkbutton_unicode")
        self.init_checkbox("org.mate.interface", "buttons-have-icons", "checkbutton_button_icons")
        
        iconSizes = Gtk.ListStore(str, str)
        iconSizes.append([_("Small"), "small-toolbar"])
        iconSizes.append([_("Large"), "large-toolbar"])
        self.builder.get_object("combobox_icon_size").set_model(iconSizes)
        self.init_combobox("org.mate.interface", "toolbar-icons-size", "combobox_icon_size")

        # Window control button
        layouts = Gtk.ListStore(str, str)
        layouts.append([_("Traditional (Right)"), "menu:minimize,maximize,close"])
        layouts.append([_("Comtemporary (Left)"), "close,minimize,maximize:"])
        self.builder.get_object("combo_wmlayout").set_model(layouts)
        self.init_combobox("org.mate.Marco.general", "button-layout", "combo_wmlayout")

        # Window manager
        wms = Gtk.ListStore(str, str)
        wms.append([_("Marco (Simple desktop effects)"), "marco"])
        if self.find_on_path('compiz'):
            wms.append([_("Compiz (Advanced GPU accelerated desktop effects)"), "compiz"])
        self.builder.get_object("combo_wm").set_model(wms)
        self.builder.get_object("combo_wm").set_tooltip_text(_("The new window manager will be activated upon selection."))
        self.init_combobox("org.mate.session.required-components", "windowmanager", "combo_wm")

        # Panel layouts
        indicators_available = False
        if os.path.exists('/usr/lib/indicators/7/libapplication.so') \
           and os.path.exists('/usr/lib/indicator-sound-gtk2/indicator-sound-service') \
           and os.path.exists('/usr/lib/mate-indicator-applet/mate-indicator-applet'):
               indicators_available = True

        mate_menu_available = False
        if os.path.exists('/usr/lib/ubuntu-mate/mate-menu/mate-menu.py'):
			mate_menu_available = True

        mint_menu_available = False
        if os.path.exists('usr/lib/linuxmint/mintMenu/mintMenu.py'):
			mint_menu_available = True			
			
        gnome_menu_available = False
        if os.path.exists('/usr/lib/gnome-main-menu/main-menu'):
            gnome_menu_available = True			

        dock = False
        if os.path.exists('/usr/bin/docky') or os.path.exists('/usr/bin/plank'):
            dock = True

        panels = Gtk.ListStore(str, str)
        if self.panel_layout_exists('default'):
            panels.append([_("MATE Desktop"), "default"])

        if self.panel_layout_exists('ubuntu-mate'):
            panels.append([_("Ubuntu MATE"), "ubuntu-mate"])

        if self.panel_layout_exists('ubuntu-mate-fresh') \
            and mate_menu_available:
            panels.append([_("Ubuntu MATE with MATE Menu"), "ubuntu-mate-fresh"])

        if self.panel_layout_exists('ubuntu-mate-indicators') \
            and indicators_available:
            panels.append([_("Ubuntu MATE with Indicators"), "ubuntu-mate-indicators"])

        if self.panel_layout_exists('ubuntu-mate-indicators-fresh') \
            and indicators_available \
            and mate_menu_available:
            panels.append([_("Ubuntu MATE with Indicators and MATE Menu"), "ubuntu-mate-indicators-fresh"])

        if self.panel_layout_exists('redmond'):
            panels.append([_("Redmond"), "redmond"])

        if self.panel_layout_exists('redmond-fresh') \
            and mate_menu_available:
            panels.append([_("Redmond with MATE Menu"), "redmond-fresh"])

        if self.panel_layout_exists('redmond-indicators') \
            and indicators_available:
            panels.append([_("Redmond with Indicators"), "redmond-indicators"])

        if self.panel_layout_exists('redmond-indicators-fresh') \
            and indicators_available \
            and mate_menu_available:
            panels.append([_("Redmond with Indicators and MATE Menu"), "redmond-indicators-fresh"])
        
        if dock:    
			if self.panel_layout_exists('eleven'):
				panels.append([_("Eleven"), "eleven"])

			if self.panel_layout_exists('eleven-fresh') \
				and mate_menu_available:
				panels.append([_("Eleven with MATE Menu"), "eleven-fresh"])

			if self.panel_layout_exists('eleven-indicators') \
				and indicators_available:
				panels.append([_("Eleven with Indicators"), "eleven-indicators"])

			if self.panel_layout_exists('eleven-indicators-fresh') \
				and indicators_available \
				and mate_menu_available:
				panels.append([_("Eleven with Indicators and MATE Menu"), "eleven-indicators-fresh"])            

        if self.panel_layout_exists('linuxmint') \
            and mint_menu_available:
            panels.append([_("Linux Mint"), "linuxmint"])

        if self.panel_layout_exists('opensuse') \
            and gnome_menu_available:
            panels.append([_("openSUSE"), "opensuse"])        

        self.builder.get_object("combobox_panels").set_model(panels)
        self.builder.get_object("combobox_panels").set_tooltip_text(_("The new panel layout will be activated on selection and destroy any customisations you might have made."))
        self.init_combobox("org.mate.panel", "default-layout", "combobox_panels")

        # toolbar icon styles
        iconStyles = Gtk.ListStore(str, str)
        iconStyles.append([_("Text below items"), "both"])
        iconStyles.append([_("Text beside items"), "both-horiz"])
        iconStyles.append([_("Icons only"), "icons"])
        iconStyles.append([_("Text only"), "text"])
        self.builder.get_object("combobox_toolicons").set_model(iconStyles)
        self.init_combobox("org.mate.interface", "toolbar-style", "combobox_toolicons")
        self.builder.get_object("main_window").show()

    
if __name__ == "__main__":
    MateTweak()
    Gtk.main()
